### PR TITLE
#649: multiple language support

### PR DESCRIPTION
**Pull request by oznasi1**
Added infra to support multiple languages using i18next library - https://www.i18next.com/.
That POC contains the option to change language between English (default) and Spanish using the burger menu.

Also, we are supporting detecting user language, load it, store it in local storage and if the user change the language manually, local storage is updating and in the next entrance to the app, we will get the language from there.

For now, I'm using i18next Backend which allows us to import on demand only the language that we use and not all of them at ones. that's why I wrapped the app with Suspense.

Also, we have a cache based on local storage with versioning and expiration time of 7 days. The flow is: 
App is loaded, check local storage available with same version - getting translation from there. if local storage does not exist so we get the translation and update local storage. And of course if version is not updated we are getting the translation from translation file and update local storage.

Also, for each translation key, Iv'e added default value to protects us in case of missing key or some other unknown error.

### CODE REVIEW

File: public/locales/en-US/translation.json [hunk 1]
```diff
@@ -0,0 +1,20 @@
+{
+  "common": {
```
Comment (pingustar) [inline public/locales/en-US/translation.json:2]: I wouldnt use common attributes for the translations. 

I think you will run into unwanted side effects when changing text that does not apply to all "common" occurences.

File: public/locales/en-US/translation.json [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    "cancelButton": "Cancel"
+  },
+  "modal": {
+    "delete": {
```
Comment (pingustar) [inline public/locales/en-US/translation.json:6]: modal.delete is not very descriptive. What are we deleting? Currently we have only strategies but lets prepare for other possible features in the future where we would have another delete modal.

File: src/components/core/menu/mainMenu/MainMenuRightBurger/useBurgerMenuItems.tsx [hunk 3]
```diff
@@ -8,6 +8,8 @@ import { ReactComponent as IconTelegram } from 'assets/logos/telegram.svg';
 import { ReactComponent as IconV } from 'assets/icons/v.svg';
 import { useFiatCurrency } from 'hooks/useFiatCurrency';
 import { MenuItemActions } from './useMenuContext';
+import { useTranslation } from 'libs/translations';
+import { SUPPORTED_LANGUAGES } from 'languages/i18n';
```
Comment (pingustar) [inline src/components/core/menu/mainMenu/MainMenuRightBurger/useBurgerMenuItems.tsx:12]: I would expect this file to be in the `libs/translations` folder

File: src/index.tsx [hunk 4]
```diff
@@ -1,17 +1,19 @@
...
+import { App } from 'App';
+import { StoreProvider } from 'store';
 import { Web3ReactWrapper } from 'libs/web3';
 import { Router } from 'libs/routing';
 import { LazyMotion } from 'libs/motion';
 import { QueryProvider } from 'libs/queries';
-import { StoreProvider } from 'store';
+import { CarbonLogoLoading } from 'components/common/CarbonLogoLoading';
+import './languages/i18n';
+import 'global-shim';
```
Comment (pingustar) [inline src/index.tsx:12]: why are you reordering these imports? I am pretty sure there was a reason that global shim and sentry is at the top.

File: src/index.tsx [hunk 5]
```diff
@@ -24,7 +26,17 @@ root.render(
         <Web3ReactWrapper>
           <LazyMotion>
             <Router>
-              <App />
+              <Suspense
```
Comment (pingustar) [inline src/index.tsx:29]: why do we introduce suspense for i18n support? Is it required/related?

File: src/libs/modals/modals/ModalConfirmStrategy/ModalConfirmStrategy.tsx [hunk 6]
```diff
@@ -63,7 +65,7 @@ export const ModalConfirmStrategy: ModalFC<ModalConfirmStrategyData> = ({
         <IconTitleText
           variant={data?.variant}
           icon={data?.icon}
-          title={data?.title || ''}
+          title={data?.subtitle || ''}
```
Comment (pingustar) [inline src/libs/modals/modals/ModalConfirmStrategy/ModalConfirmStrategy.tsx:68]: is this intentional? 

title vs. subtitle

File: src/libs/translations/index.ts [hunk 7]
```diff
@@ -0,0 +1,8 @@
+import HttpBackend from 'i18next-http-backend';
```
Comment (pingustar) [inline src/libs/translations/index.ts:1]: cant you export directly?

`export HttpBackend from "i18next-http-backend"`

Comment (oznasi1) [reply src/libs/translations/index.ts:1]: Yes, wanted to import as default and name export so I used that:
export { default as HttpBackend } from 'i18next-http-backend';

Applied to all imports in the file.

File: src/libs/modals/modals/ModalConfirmStrategy/ModalConfirmStrategy.tsx (hunk 6 above)
Comment (oznasi1) [reply src/libs/modals/modals/ModalConfirmStrategy/ModalConfirmStrategy.tsx:68]: it was, but renamed it now to be more clear.

File: src/components/core/menu/mainMenu/MainMenuRightBurger/useBurgerMenuItems.tsx [hunk 8]
```diff
@@ -174,10 +180,33 @@ export const useBurgerMenuItems = () => {
     []
   );
 
-  menuMap.set('main', { items: mainItems });
+  const languagesItems = useMemo(
+    (): MenuItemType[] =>
+      SUPPORTED_LANGUAGES.map(({ code, name }) => {
```
[the remark this answers is not in the document]
Comment (oznasi1) [reply src/components/core/menu/mainMenu/MainMenuRightBurger/useBurgerMenuItems.tsx:185]: @pingustar  Do you think the right way is to split all menu items to their own hook?

[the remark this answers is not in the document]
Comment (oznasi1) [reply src/components/core/menu/mainMenu/MainMenuRightBurger/useBurgerMenuItems.tsx:185]: @pingustar After discussion with Jan, that is good for now but be aware for further additional to that file any consider to split this file out.

File: .eslintrc.json [hunk 9]
```diff
@@ -0,0 +1,14 @@
+{
```
Comment (zavelevsky) [inline .eslintrc.json:1]: please move all eslint config here from the package.json

File: src/components/common/forwardArrow.tsx [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+export const ForwardArrow: FC<ForwardArrowProps> = ({
+  arrowType = 'cut',
+  className = '',
+}) => {
+  const { i18n } = useTranslation();
+  const newClassName = `h-12 w-7 ${
+    i18n.dir() === 'rtl' ? 'rotate-180' : ''
+  } ${className}`;
+
+  return (
```
Comment (zavelevsky) [inline src/components/common/forwardArrow.tsx:20]: why not 
```js
return arrowType === 'cut' ? <IconArrowCut className={newClassName} /> : <IconArrowFull className={newClassName} />
```

File: src/components/common/walletConnect/utils.tsx [hunk 11]
```diff
@@ -0,0 +1,38 @@
+import { ReactComponent as IconWallet1 } from 'assets/icons/wallet-1.svg';
+import { ReactComponent as IconWallet2 } from 'assets/icons/wallet-2.svg';
+import { ReactComponent as IconWallet3 } from 'assets/icons/wallet-3.svg';
+import { ReactComponent as IconWallet4 } from 'assets/icons/wallet-4.svg';
+import { ReactComponent as IconWallet5 } from 'assets/icons/wallet-5.svg';
+import { ReactComponent as IconWallet6 } from 'assets/icons/wallet-6.svg';
+import { i18n } from 'libs/translations';
+
+export const getProductDescriptionItems = () => {
```
Comment (zavelevsky) [inline src/components/common/walletConnect/utils.tsx:9]: utils isn't a good name for this file

Comment (oznasi1) [reply src/components/common/walletConnect/utils.tsx:9]: amm, is data.tsx sounds better ?

Comment (oznasi1) [reply src/components/common/walletConnect/utils.tsx:9]: @zavelevsky can u take a look at this again pls?

File: src/components/common/TokenInputField/TokenInputField.tsx [hunk 12]
```diff
@@ -33,11 +34,12 @@ export const TokenInputField: FC<Props> = ({
   isError,
   className,
   onKeystroke,
-  placeholder = 'Enter Amount',
+  placeholder = i18n.t('common.placeholders.placeholder2'),
```
Comment (zavelevsky) [inline src/components/common/TokenInputField/TokenInputField.tsx:37]: so why use the hook at all?

Comment (oznasi1) [reply src/components/common/TokenInputField/TokenInputField.tsx]: As we talked, we are now using t function from hook only and pass it down to functions if needed.

File: src/components/common/walletConnect/utils.tsx (hunk 11 above)
Comment (oznasi1) [reply src/components/common/walletConnect/utils.tsx]: Changed to items.tsx

File: src/components/core/menu/mainMenu/MainMenuRightBurger/useBurgerMenuItems.tsx [hunk 13]
```diff
@@ -25,87 +27,96 @@ const iconStyles = 'h-32 w-32 md:h-20 md:w-20';
...
-      },
-      {
-        subMenu: 'resources',
-        content: 'Resources',
-      },
-      {
-        content: (
-          <Link className="flex" to={externalLinks.faq}>
-            FAQ
+  const mainItems: MenuItemType[] = [
```
Comment (zavelevsky) [inline src/components/core/menu/mainMenu/MainMenuRightBurger/useBurgerMenuItems.tsx:36]: you removed useMemo in a couple of places - why?

File: src/libs/modals/modals/ModalGenericInfo.tsx [hunk 14]
```diff
@@ -52,15 +53,15 @@ export const ModalGenericInfo: ModalFC<ModalGenericInfoData> = ({
         }}
         className={'my-16'}
       >
-        {buttonLabel}
+        {buttonLabel || t('modals.genericInfo.actionButtons.actionButton1')}
```
Comment (zavelevsky) [inline src/libs/modals/modals/ModalGenericInfo.tsx:56]: note that you don't allow me to pass an empty string - becuase this check will always override it. You do it in other places too. Not a big issue - but still.

**Review by Costa769 — APPROVED** [review]

**Comment by ashachaf** [comment]
to support language on analytics:
change the way we pass page referral back to analytics.
current:
https://github.com/search?q=repo%3Abancorprotocol%2Fcarbon-app%20page_referrer_spa&type=code

updated:
```
page:
 {
    referrer_spa: "/about", //where "/about" is dynamic based on the viewed page
    language: "en"
  }
```

**Pull request merged**